Draft : Add option to lock/unlock window resizing via keyboard shortcut

TusharGautam29 opened the pull request:
A new feature that allows users to toggle window resizing on or off that users can toggle it dynamically via a keyboard shortcut.(with F11)

Feature- #15124
[#15124](https://bugs.scummvm.org/ticket/15124)

Changing the flag of window created based on the private variable _lockedScreen.  Will add a SetLockedScreen(bool val) function and implement it into event system.

Request for Feedback

bluegr commented on backends/platform/sdl/win32/win32-main.cpp:
```diff
@@ -43,7 +43,7 @@ int __stdcall WinMain(HINSTANCE /*hInst*/, HINSTANCE /*hPrevInst*/,  LPSTR /*lpC
 #endif
 // HACK: __argc, __argv are broken and return zero when using mingwrt 4.0+ on MinGW
-// HACK: MinGW-w64 based toolchains neither feature _argc nor _argv. The 32 bit
+//- HACK: MinGW-w64 based toolchains neither feature _argc nor _argv. The 32 bit
```
This seems to have been modified by mistake

```suggestion
// HACK: MinGW-w64 based toolchains neither feature _argc nor _argv. The 32 bit
```

bluegr commented:
Looks good at a first glance, up to now
Please adapt the commit message according to our [commit guidelines](https://wiki.scummvm.org/index.php?title=Commit_Guidelines)

lephilousophe commented:
The 3D graphics part (in graphics3d/openglsdl) is also missing.

TusharGautam29 commented:
While working on this feature i noticed that if I initialise _lockedScreen = true then no GUI theme can load, 
Error - Failed to load any GUI theme, aborting!

I removed all the _lockedScreen variables , now only one exists in WindowedGraphicsManager.

OpenGLSdlGraphics3dManager, OpenGLGraphicsManager ,SdlGraphicsManager now has a function to toggle the variable.

I'll commit the code so you can analyse and help me with it

lephilousophe commented on backends/graphics/windowed.h:
```diff
@@ -115,6 +115,10 @@ class WindowedGraphicsManager : virtual public GraphicsManager {
 	int getWindowHeight() const { return _windowHeight; }
 
+	// New methods to lock and unlock the screen
+	virtual void setLockedScreen(bool val) { _lockedScreen = val; }
```
Why do you define getter and setter?
You could put the variable protected and it would be accessible to children classes.

lephilousophe commented on backends/graphics/opengl/opengl-graphics.cpp:
```diff
@@ -1272,57 +1275,67 @@ void OpenGLGraphicsManager::grabPalette(byte *colors, uint start, uint num) cons
 void OpenGLGraphicsManager::handleResizeImpl(const int width, const int height) {
 	// Setup backbuffer size.
-	_targetBuffer->setSize(width, height, getRotationMode());
+	if (WindowedGraphicsManager::isScreenLocked()) {
```
What's the purpose of this?

lephilousophe commented on backends/graphics/opengl/opengl-graphics.cpp:
```diff
@@ -207,6 +207,9 @@ bool OpenGLGraphicsManager::setGraphicsMode(int mode, uint flags) {
 int OpenGLGraphicsManager::getGraphicsMode() const {
 	return _currentState.graphicsMode;
 }
+void OpenGLGraphicsManager::setLockedScreen(bool val) {
```
This is completely unnecessary.

lephilousophe commented on backends/graphics/openglsdl/openglsdl-graphics.cpp:
```diff
@@ -299,6 +299,9 @@ void OpenGLSdlGraphicsManager::initSize(uint w, uint h, const Graphics::PixelFor
 
 	return OpenGLGraphicsManager::initSize(w, h, format);
 }
+void OpenGLSdlGraphicsManager::setLockedScreen(bool val) {
```
Either you put stuff in WindowedGraphicsManager, either you put it in SdlGraphicsManager.
In both case this is useless.

lephilousophe commented on backends/graphics/sdl/sdl-graphics.cpp:
```diff
@@ -455,7 +458,9 @@ bool SdlGraphicsManager::notifyEvent(const Common::Event &event) {
 	}
 }
-
+void SdlGraphicsManager::setLockedScreen(bool val) {
```
Same, useless. And it misses empty lines around.

lephilousophe commented on backends/graphics/surfacesdl/surfacesdl-graphics.cpp:
```diff
@@ -1554,6 +1554,10 @@ void SurfaceSdlGraphicsManager::internUpdateScreen() {
 #endif
 }
 
+void SurfaceSdlGraphicsManager::setLockedScreen(bool val) {
```
To be removed.

lephilousophe reviewed (changes requested):
Do you really need to touch to WindowedGraphicsManager and OpenGLGraphicsManager?

These classes are used by many other backends and touching them doesn't seem necessary to me.
The first commit seemed to start better.

Please note that SdlGraphicsManager is a common ancestor for all SDL graphical backends (2D and 3D).

TusharGautam29 commented:
Ok, i'll look into the changes you suggested and work as per that. Thank you

TusharGautam29 commented:
I removed it from  WindowedGraphicsManager as it was causing issues if i set it to true, This is working for both cases, Also i added it to event system, but doesn't seem to be working. Any feedback on that please

sev- commented on backends/graphics/opengl/opengl-graphics.cpp:
```diff
@@ -1294,16 +1294,15 @@ void OpenGLGraphicsManager::handleResizeImpl(const int width, const int height)
 	// anyway. Thus, it should not be a real issue for modern hardware.
-	if (   overlayWidth  > (uint)OpenGLContext.maxTextureSize
-	    || overlayHeight > (uint)OpenGLContext.maxTextureSize) {
+	if (overlayWidth > (uint)OpenGLContext.maxTextureSize || overlayHeight > (uint)OpenGLContext.maxTextureSize) {
```
These changes are irrelevant, please remove all of them. If you want/need to do whitespace changes, those should go separately, but in this particular case, they are made on purpose, and obviously, it is just your code editor reformatting them, thus, please revert.

sev- commented on backends/graphics/openglsdl/openglsdl-graphics.cpp:
```diff
@@ -299,7 +299,6 @@ void OpenGLSdlGraphicsManager::initSize(uint w, uint h, const Graphics::PixelFor
 
 	return OpenGLGraphicsManager::initSize(w, h, format);
 }
-
```
Please revert

sev- commented on backends/graphics/openglsdl/openglsdl-graphics.cpp:
```diff
@@ -914,6 +945,7 @@ bool OpenGLSdlGraphicsManager::notifyEvent(const Common::Event &event) {
 		return true;
 	}
 
+
```
Please revert

sev- commented on backends/graphics/openglsdl/openglsdl-graphics.h:
```diff
@@ -97,6 +101,8 @@ class OpenGLSdlGraphicsManager : public OpenGL::OpenGLGraphicsManager, public Sd
 	uint _graphicsScale;
 	bool _gotResize;
 
+	bool _lockedScreen = false; // New member variable to track lock state
```
What makes it "New" besides being in this PR? :)
```suggestion
	bool _lockedScreen = false; // Variable to track lock state
```

sev- commented on backends/graphics/sdl/sdl-graphics.cpp:
```diff
@@ -455,7 +472,6 @@ bool SdlGraphicsManager::notifyEvent(const Common::Event &event) {
 		return false;
 	}
 }
-
```
Same thing. Please review all files and revert all of these whitespace changes. I will not comment on them more, just please revert.

sev- commented on backends/graphics/windowed.h:
```diff
@@ -65,7 +65,7 @@ class WindowedGraphicsManager : virtual public GraphicsManager {
 		_cursorNeedsRedraw(false),
 		_cursorLastInActiveArea(true) {}
-
+		
```
This is even worse: it adds a trailing tab. Please revert

sev- reviewed (changes requested):
Added some notes. Please cleanup.

lephilousophe commented on backends/graphics/opengl/opengl-graphics.cpp:
```diff
@@ -1294,16 +1294,16 @@ void OpenGLGraphicsManager::handleResizeImpl(const int width, const int height)
 	// anyway. Thus, it should not be a real issue for modern hardware.
-	if (   overlayWidth  > (uint)OpenGLContext.maxTextureSize
-	    || overlayHeight > (uint)OpenGLContext.maxTextureSize) {
+	if (overlayWidth > (uint)OpenGLContext.maxTextureSize
```
This should not have been touched

lephilousophe commented on backends/graphics/openglsdl/openglsdl-graphics.h:
```diff
@@ -97,6 +101,8 @@ class OpenGLSdlGraphicsManager : public OpenGL::OpenGLGraphicsManager, public Sd
 	uint _graphicsScale;
 	bool _gotResize;
 
+	bool _lockedScreen = false; // Variable to track lock state
```
This variable name is not correct: what is locked is the windows size, nothing else.
The comment is useless.
Something like `_resizableWindow`  would fit better.

bluegr replied:
Furthermore, you don't need a variable to track the lock state - SDL already holds that info.

You could do something like:

```
bool SdlGraphicsManager::isWindowResizable() const {
    return SDL_GetWindowFlags(_window->getSDLWindow()) & SDL_WINDOW_RESIZABLE;
}
```

lephilousophe commented on backends/graphics/openglsdl/openglsdl-graphics.cpp:
```diff
@@ -362,8 +385,8 @@ void OpenGLSdlGraphicsManager::notifyResize(const int width, const int height) {
 	}
-
-	handleResize(currentWidth, currentHeight);
+	if (!_lockedScreen)
```
This doesn't look good.
Even if the window is not resizable, it could be resized by the window manager.

lephilousophe commented on backends/graphics/openglsdl/openglsdl-graphics.cpp:
```diff
@@ -335,6 +335,29 @@ void OpenGLSdlGraphicsManager::updateScreen() {
 	OpenGLGraphicsManager::updateScreen();
 }
 
+void OpenGLSdlGraphicsManager::lockWindowReSize() {
```
These two functions (lock and unlock) don't bring much as you are only calling them at one place.

lephilousophe commented on backends/graphics/openglsdl/openglsdl-graphics.cpp:
```diff
@@ -718,7 +744,13 @@ bool OpenGLSdlGraphicsManager::notifyEvent(const Common::Event &event) {
 		return SdlGraphicsManager::notifyEvent(event);
 	}
-
+	if (event.type == Common::EVENT_KEYDOWN) {
```
This is not the way to do this.
Please look at how fullscreen switching is done.

I also bet this code has not been tested on with any other renderer than the 2D OpenGL one.
I already suggested you that SdlGraphicsManager is a common ancestor for all SDL graphical backends (2D and 3D).

lephilousophe commented on backends/graphics/surfacesdl/surfacesdl-graphics.h:
```diff
@@ -449,6 +449,8 @@ class SurfaceSdlGraphicsManager : public SdlGraphicsManager {
 	void setFullscreenMode(bool enable);
 	void handleScalerHotkeys(uint mode, int factor);
 
+	bool _lockedScreen = false;
```
You don't seem to understand how inheritance work.

lephilousophe commented on engines/bagel/boflib/gui/window.cpp:
```diff
@@ -257,6 +257,8 @@ void CBofWindow::move(const int x, const int y, bool bRepaint) {
 }
 
 void CBofWindow::reSize(CBofRect *pRect, bool bRepaint) {
+	if (_lockedScreen)
```
Why is this even touched?

lephilousophe reviewed (commented):
I have made a new review.

I wonder what's your objective with this PR.
Is this solely to implement this feature or do you have some other plans?

bluegr commented on backends/graphics/sdl/sdl-graphics.cpp:
```diff
@@ -336,6 +347,12 @@ bool SdlGraphicsManager::createOrUpdateWindow(int width, int height, const Uint3
 	if (!_window) {
 		return false;
 	}
+	Uint32 windowFlags = flags;
+	if (_lockedScreen) {
+		windowFlags &= ~SDL_WINDOW_RESIZABLE;
+	} else {
+		windowFlags |= SDL_WINDOW_RESIZABLE;
+	}
```
Since `windowFlags` is not used anywhere, this code doesn't affect any functionality

bluegr commented on backends/graphics/opengl/opengl-graphics.cpp:
```diff
@@ -1294,16 +1294,16 @@ void OpenGLGraphicsManager::handleResizeImpl(const int width, const int height)
 	// anyway. Thus, it should not be a real issue for modern hardware.
-	if (   overlayWidth  > (uint)OpenGLContext.maxTextureSize
-	    || overlayHeight > (uint)OpenGLContext.maxTextureSize) {
+	if (overlayWidth > (uint)OpenGLContext.maxTextureSize
+		|| overlayHeight > (uint)OpenGLContext.maxTextureSize) {
```
Please revert the whitespace changes

```suggestion
	if (   overlayWidth  > (uint)OpenGLContext.maxTextureSize
	    || overlayHeight > (uint)OpenGLContext.maxTextureSize) {
```

bluegr commented on engines/bagel/boflib/gui/window.cpp:
```diff
@@ -257,6 +257,8 @@ void CBofWindow::move(const int x, const int y, bool bRepaint) {
 }
 
 void CBofWindow::reSize(CBofRect *pRect, bool bRepaint) {
+	if (_lockedScreen)
+		return;
```
This is unnecessary and out of scope of this PR

```suggestion
```

bluegr commented on engines/bagel/boflib/gui/window.h:
```diff
@@ -51,6 +51,7 @@ class CBofWindow : public CLList, public CBofObject, public CBofError {
 private:
 	bool _visible = true;
 	bool _enabled = true;
+	bool _lockedScreen = true;
```
```suggestion
```

bluegr commented:
Thanks for working on this. This has been superseded by PR #6449. Closing
